tests/wlr_layer_shell_v1.cpp

```diff
@@ -82,119 +90,142 @@ class LayerSurfaceTest: public wlcs::StartedInProcessServer
-                output_x :
-                (right ?
-                    (output_x + output_width - width) :
+            (anchor.left ?
+                output_x + margin.left :
+                (anchor.right ?
+                    (output_x + output_width - width - margin.right) :
                     (output_x + (output_width - width) / 2)
                 )
             );
         int const y =
-            (top ?
-                output_y :
-                (bottom ?
-                    (output_y + output_height - height) :
+            (anchor.top ?
+                output_y + margin.top :
+                (anchor.bottom ?
+                    (output_y + output_height - height - margin.bottom) :
                     (output_y + (output_height - height) / 2)
                 )
             );
         return std::make_pair(std::make_pair(x, y), std::make_pair(width, height));;
     }
 
+    auto request_size() const -> Vec2
+    {
+        return std::make_pair(
+            h_expand() ? 0 : LayerSurfaceTest::default_width,
+            v_expand() ? 0 : LayerSurfaceTest::default_height);
+    }
+
     auto configure_size(Rect const& output) const -> Vec2
     {
-        int const configure_width = h_expand() ? output.second.first : 0;
-        int const configure_height = v_expand() ? output.second.second : 0;
+        int const configure_width = h_expand() ? output.second.first - margin.left - margin.right : 0;
+        int const configure_height = v_expand() ? output.second.second - margin.top - margin.bottom : 0;
         return std::make_pair(configure_width, configure_height);
     }
 
     // Will always either return 0, or a single enum value
     auto attached_edge() const -> zwlr_layer_surface_v1_anchor
     {
-        if (top == bottom)
+        if (anchor.top == anchor.bottom)
         {
-            if (left && !right)
+            if (anchor.left && !anchor.right)
                 return ZWLR_LAYER_SURFACE_V1_ANCHOR_LEFT;
-            else if (right && !left)
+            else if (anchor.right && !anchor.left)
                 return ZWLR_LAYER_SURFACE_V1_ANCHOR_RIGHT;
         }
-        else if (left == right)
+        else if (anchor.left == anchor.right)
         {
-            if (top && !bottom)
+            if (anchor.top && !anchor.bottom)
                 return ZWLR_LAYER_SURFACE_V1_ANCHOR_TOP;
-            else if (bottom && !top)
+            else if (anchor.bottom && !anchor.top)
                 return ZWLR_LAYER_SURFACE_V1_ANCHOR_BOTTOM;
         }
         return (zwlr_layer_surface_v1_anchor)0;
     }
 
-    bool const left, right, top, bottom;
+    Sides<bool> const anchor;
+    Sides<int> const margin;
 };
 
-std::ostream& operator<<(std::ostream& os, const LayerAnchor& anchor)
+// Given a LayerSurfaceLayout, emits the margin sides to use in a call to zwlr_layer_surface_v1_set_margin()
+#define SPLAT_MARGIN(layout) (layout).margin.top, (layout).margin.right, (layout).margin.bottom, (layout).margin.left
```

Comment:
I am not wild about `SPLAT_MARGIN`; I see what you're doing with it, but I don't think the small decrease in verbosity is worth the surprise of a macro expanding into multiple arguments.

Reply:
This is the problem with code review, you can't get anything fun by it